Feature: Show/hide helper Nutrition field depending on whether the information is collected for a Job

NiklasRosenstein opened the pull request:
This PR adds the requested functionality that a user is not prompted for a Nutrition preference when none of the shifts they are looking to sign up for provide food to the helper.

### UX changes

* The nutrition field in the helper registration form is now hidden if `Event.ask_nutrition` is false, or if the user has not selected any shifts for a Job where `ask_nutrition` is set to true. The default value for nutrition is saved in the helper registration regardless.

  <table><tr><th>With nutrition</th><th>Without nutrition</th></tr><tr><td><img width="1340" alt="image" src="https://user-images.githubusercontent.com/1318438/232254831-02926285-9d56-473f-9aa7-48da4242ac57.png"></td><td><img width="1337" alt="image" src="https://user-images.githubusercontent.com/1318438/232254821-01c012c3-4449-4f3c-94af-95918b4d071a.png"></td></tr></table>


* Respectively, the field is also hidden when a helper views and edits their personal information.

  <table><tr><th>With nutrition</th><th>Without nutrition</th></tr><tr><td><img width="691" alt="image" src="https://user-images.githubusercontent.com/1318438/232254949-cbaadb49-7e11-4159-9a38-e8414e82b871.png"></td><td><img width="717" alt="image" src="https://user-images.githubusercontent.com/1318438/232254911-7187d8bf-6628-409b-a0b3-e83dffa0e264.png"></td></tr></table>

### Additional changes

* Jobs now show a "Private" badge. Before this change, only public jobs had a "Public" badge.

  <img width="286" alt="image" src="https://user-images.githubusercontent.com/1318438/232254713-5ae0b3a5-cbf0-4d43-b902-821d09c869f5.png">

### Discussion points

* [ ] If an admin enables nutrition in a Job after helpers have already registered to shifts, do we need to inform helpers that they should now select a nutrition preference?
* [ ] I remember @Tomatower mentioned something about deduplicating multiple helper registrations; should we add special logic for the nutrition there?
* [ ] Should we display a badge next to the job name that asking for nutrition is enabled (at the same place where we show the "Public" and "Private" badge?)

NiklasRosenstein commented:
@Tomatower I opted for `ask_nutrition` as the field name instead of `foodinfo_enabled`.

svenhertle commented:
Thanks for the pull request and detailed description!

I'm sorting some things right now (for example a Celery update required for python 3.11) and will merge it then. I've already done one rebase of your code.

svenhertle commented:
I've changed a few things now and it seems to work. But there are some open issues:
* If no choice was made by the helper (= they were not asked), no option should be stored. So we need to make it optional and require it on demand (similar to the infection instruction)
* Merging of helpers needs to be updated
* The Excel export maybe needs to be updated
* The statistics page needs to be updated
* We should show a warning that disabling it on a job level also means that the coordinators are not asked for their preference. This could be relevant for some events.

NiklasRosenstein commented on src/registration/models/helper.py:
```diff
@@ -354,8 +354,10 @@ def all_jobs(self):
-        """Returns True if the nutritional prefernce is required."""
+    def ask_nutrition(self):
+        """Returns True if the nutritional preference is required."""
+        if hasattr(self, "event") and not self.event.ask_nutrition:
```
When would `self.event` not be set? Seems a bit suboptimal to rely on `hasattr()` here.

NiklasRosenstein commented on src/registration/static/registration/js/form.js:
```diff
@@ -79,41 +79,13 @@ $('input.registration_possible').each(function (i, element) {
-{
-    var show_field = 0;
+function handle_optional_field(name) {
+    let show_field = 0;
```
nit: use `false` instead of `0`?

NiklasRosenstein commented on src/registration/static/registration/js/form.js:
```diff
@@ -79,41 +79,13 @@ $('input.registration_possible').each(function (i, element) {
+    $("input." + name).each(function() {
         if($(this).prop('checked'))
         {
             show_field = 1;
```
nit: use `true` instead of `1`?

NiklasRosenstein commented:
Slightly unrelated to this feature; but I added a "Private" badge to private jobs for clarity in 52409bc452e39fe01dee862763db01615947fd2c.

NiklasRosenstein commented:
> If no choice was made by the helper (= they were not asked), no option should be stored. So we need to make it optional and require it on demand (similar to the infection instruction)

If I see it correctly, the infection instructions can be blank in the model, so I'll make the nutrition field blank-able as well. a8f1bc6f

> The Excel export maybe needs to be updated

The CSV export looks OK as it is, there's no special handling for the `None` value of `infection_instructions` as well. There does appear to be some code for Excel as well but I couldn't find any mention of the nutrition field there.

> The statistics page needs to be updated

I propose we show the number of helpers that didn't specify in the pie chart if there are any. This is relevant if an event or job is later updated to ask for nutrition but helpers have already registered. I've updated the `NutritionData` to have a `num_blank` attribute, and also updated the pie chart code to actually use the class instead of duplicating the logic. a8f1bc6f

> We should show a warning that disabling it on a job level also means that the coordinators are not asked for their preference. This could be relevant for some events.

Do you suggest we do this on the client side? Alternatively, we could show a flash message/banner after enabling to ask for nutrition for an event/job. Or we show a banner to inform event admins if there are any helpers that should specify their nutrition preferences, but haven't. Clicking the banner could take them to some kind of diagnostics page.

> Merging of helpers needs to be updated

I'm not sure I understand the current deduplication logic. It doesn't seem to take into account what to do with conflicting choices about infection instructions, for example, either.
